### PR TITLE
Command registration rewrite

### PR DESCRIPTION
**Pull request by BobDotCom**
## Summary

Rewrites command registration to actually be efficient and smart.

### Needs testing
 - [x] Command groups

*And maybe more...*

### Unfinished
 - [x] Command permissions (v1)
 - [ ] Command permisisons (v2)
 - [x] Documentation (New methods)
 - [x] Documentation (Changed methods)
 - [x] Documentation (Guides/examples if applicable)
 
*And maybe more...*

If there's anything I need to add just let me know.
 
If anyone would like to contribute, just make a PR on the `feature/command-registration` branch.

### CODE REVIEW

**Comment by BobDotCom** [comment]
# Checklist
## Needs testing
 - [ ] Command groups

*And maybe more...*

## Unfinished
 - [ ] Command permissions (v1)
 - [ ] Command permisisons (v2)
 
*And maybe more...*

If there's anything I need to add just let me know.

File: discord/bot.py [hunk 1]
```diff
@@ -191,270 +198,583 @@ def get_application_command(
...
+                    for opt in to_check[check]:
+
+                        cmd_vals = [val.get(opt, MISSING) for val in as_dict[check]]
+                        for i, val in enumerate(cmd_vals):
+                            # We need to do some falsy conversion here
+                            # The API considers False (autocomplete) and [] (choices) to be falsy values
+                            falsy_vals = (False, [])
+                            if val in falsy_vals:
+                                cmd_vals[i] = MISSING
+                        match_vals = [val.get(opt, MISSING) for val in match[check]]
```
Comment (Luc1412) [inline discord/bot.py:282]: I got a `KeyError: 'options'` here. With `{'id': '938247032953511967', 'application_id': '443303689411887104', 'version': '938247032953511968', 'default_permission': False, 'default_member_permissions': None, 'type': 1, 'name': 'shutdown', 'description': 'Shutdown the bot.', 'dm_permission': None}` as `match`

File: discord/bot.py [hunk 2]
```diff
@@ -191,270 +198,583 @@ def get_application_command(
...
+            Registers the commands regardless of the state of the command on discord, this can take up more API calls
+            but is sometimes a more foolproof method of registering commands. This also sometimes causes minor bugs
+            where the command can temporarily appear as an invalid command on the user's side. Defaults to False.
         """
-        commands_to_bulk = []
+        if commands is None:
+            commands = self.pending_application_commands
 
-        needs_bulk = False
+        commands = copy.deepcopy(commands)
```
Comment (Luc1412) [inline discord/bot.py:372]: When using deepcopy with a list of commands it raises `TypeError: cannot pickle 'generator' object`.

**Comment by Luc1412** [comment]
Generally it currently it doesn't act as smart as it could. It still often re-creates all commands even trough I didn't modified any command. With a few restart I reached the 200 command creation limit.

An option for not deleting missing commands would be useful in case a cog isn't loaded or a command is added/updated later on runtime.


A perfect flow would be:
First startup: Bulk upsert all commands
Restart withou modifing cmds: No Action
Restart with modifing a command/updating a command on runtime: Only edit the command.

File: discord/bot.py (hunk 1 above)
Comment (BobDotCom) [reply discord/bot.py]: resolved in 905f6e4301db725cdb231890218a9df2df035fc0

File: discord/bot.py (hunk 2 above)
Comment (BobDotCom) [reply discord/bot.py:371]: sounds like you passed a generator, not a list

**Comment by BobDotCom** [comment]
> Generally it currently it doesn't act as smart as it could. It still often re-creates all commands even trough I didn't modified any command. With a few restart I reached the 200 command creation limit.
> 
> An option for not deleting missing commands would be useful in case a cog isn't loaded or a command is added/updated later on runtime.
> 
> A perfect flow would be: First startup: Bulk upsert all commands Restart withou modifing cmds: No Action Restart with modifing a command/updating a command on runtime: Only edit the command.
> 
> This also applies to multi process bots: First Process starts: Register all commands Seconds process starts: Just fetch the added commands ....

I haven't encountered this at all, are you sure? The whole reason this was rewritten was because the commands shouldn't be bulk registered every time if they don't need to be. This setup should utilize the minimum amount of registrations possible, though it will hit its limits if you start up without commands a couple times.

**Review by VincentRPS — APPROVED** [review]
lgtm

**Review by Lulalaby — DISMISSED** [review]

**Comment by BobDotCom** [comment]
Forgot to make one change

**Review by VincentRPS — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]

**Review by ChickenDevs — APPROVED** [review]
LGTM

**Pull request merged**